File: LeetCode/1639.Number_of_Ways_to_Form_a_Target_String_Given_a_Dictionary/Mario/1639.Number_of_Ways_to_Form_a_Target_String_Given_a_Dictionary.java
class Solution {
    public int numWays(String[] words, String target) {
        final int M = 1_000_000_007;
        int n = target.length(), m = words[0].length();
        int[][] counter = new int[26][m + 1];

        for (String word : words) {
            for (int i = 0; i < m; i++) {
                counter[word.charAt(i) - 'a'][i + 1] += 1;
            }
        }

        // dp[i][j] : how many ways to form target[0:i] with counter[26][k];
        // not choose jth letter: dp[i][j] = dp[i]dp[j-1]
        // choose jth letter: dp[i][j] = dp[i-1]dp[j-1] * counter[x][k]
        long[][] dp = new long[n + 1][m + 1];

        // init
        for (int i = 0; i <= m; i++)
            dp[0][i] = 1;

        target = "#" + target;

        for (int i = 1; i <= n; i++) {
            for (int j = i; j <= m; j++) {
                dp[i][j] = dp[i][j - 1];
                dp[i][j] %= M;
                if (counter[target.charAt(i) - 'a'][j] > 0) {
                    dp[i][j] += dp[i - 1][j - 1] * counter[target.charAt(i) - 'a'][j];
                    dp[i][j] %= M;
                }
            }
        }

        // for (long[] d : dp) {
        // System.out.println(Arrays.toString(d));
        // }

        return (int) dp[n][m];
    }
}

File: LeetCode/1639.Number_of_Ways_to_Form_a_Target_String_Given_a_Dictionary/Mario/1639.Number_of_Ways_to_Form_a_Target_String_Given_a_Dictionary.py
#!/usr/bin/env python3


class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        # 1. from left to right
        # 2. target[i] = words[j][k]
        # 3. used kth, can't use xth <= k in every word

        # dp[len(word)][a - z]:
        # "aba" i = 1, b
        # [1-2] to find b,
        # 1th b: [0,0] a
        # 2th b: 1th b + [1, 1] a

        MOD = 10**9 + 7
        m, n = len(words), len(words[0])
        count = [[0] * 26 for _ in range(n)]

        for w in words:
            for i, c in enumerate(w):
                j = ord(c) - ord("a")
                count[i][j] += 1

        # dp
        dp = [[0] * 26 for _ in range(n)]
        # init dp

        for i, c in enumerate(target):
            left_letter_count = len(target) - (i + 1)
            idx = ord(c) - ord("a")
            res = 0
            for j in range(n - left_letter_count - 1, i - 1, -1):
                if count[j][idx] > 0:
                    if i == 0:
                        dp[j][idx] = count[j][idx]
                    else:
                        # find previous letter, accumulate count
                        prev = ord(target[i - 1]) - ord("a")
                        prev_count = 0
                        for k in range(i - 1, j):
                            if dp[k][prev] > 0:
                                prev_count = (prev_count + dp[k][prev]) % MOD
                        dp[j][idx] = (count[j][idx] * prev_count) % MOD
                        res = (res + dp[j][idx]) % MOD

        # for r in dp:
        #     print(r)

        return res


class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        # dp[i][j] means the num of way word[:i] form target[:j]
        # dp[i][j] state transition equation
        # 1. not pick current col, dp[i][j] = dp[i][j-1]
        # 2. pick current col: dp[i][j] += dp[i-1][j-j] * count[target[j]]

        MOD = 10**9 + 7
        target = "#" + target

        m, n = len(words[0]), len(target)

        # counter
        count = [Counter() for _ in range(m)]
        for w in words:
            for i, c in enumerate(w):
                count[i][c] += 1

        # dp
        dp = [[0] * (m + 1) for _ in range(n)]
        # init dp: every col in word can form empty str
        for i in range(m + 1):
            dp[0][i] = 1

        for i in range(1, n):
            cur = target[i]
            left = n - i
            for j in range(1, m + 1):
                # not take current col
                dp[i][j] = dp[i][j - 1]
                # take current col
                if cur in count[j - 1]:
                    dp[i][j] = (dp[i][j] + dp[i - 1][j - 1] * count[j - 1][cur]) % MOD

        # for r in dp:
        #     print(r)

        return dp[n - 1][m]
